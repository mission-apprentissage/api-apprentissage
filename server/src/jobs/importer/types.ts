import type { CronDef } from "job-processor";
import type { ImportStatus } from "shared";

export type Importer = CronDef & {
  getStatus(): Promise<ImportStatus>;
};
