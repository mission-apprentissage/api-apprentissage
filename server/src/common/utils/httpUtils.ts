import axios, { AxiosInstance, AxiosRequestConfig, CreateAxiosDefaults } from "axios";
import http from "http";
import https from "https";
import iconv from "iconv-lite";
import { compose } from "oleoduc";

import logger from "@/common/logger";

// https://github.com/axios/axios/issues/3845#issuecomment-1040819908
class BufferedHttpAgent extends http.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);

    this.highWaterMark = highWaterMark;
  }
  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}
class BufferedHttpsAgent extends https.Agent {
  highWaterMark: number;

  constructor({ highWaterMark = 16 * 1024, ...rest }) {
    super(rest);
    //see https://github.com/nodejs/node/issues/39092
    this.highWaterMark = highWaterMark;
  }

  // @ts-expect-error: TODO
  createConnection(options, callback) {
    // @ts-expect-error
    return super.createConnection({ ...options, highWaterMark: this.highWaterMark }, callback);
  }
}

type FetchOptions = AxiosRequestConfig & {
  highWaterMark: number;
  raw: boolean;
  encoding: string;
};
async function _fetch(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { method = "GET", data, highWaterMark, ...rest } = options;
  logger.debug(`${method} ${url}...`);

  return client.request({
    url,
    method,
    httpAgent: new BufferedHttpAgent({ highWaterMark }),
    httpsAgent: new BufferedHttpsAgent({ highWaterMark }),
    ...(data ? { data } : {}),
    ...rest,
  });
}

async function fetchStream(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, encoding = "UTF-8", ...rest } = options;
  const response = await _fetch(url, { responseType: "stream", ...rest }, client);

  const stream = compose(response.data, iconv.decodeStream(encoding));

  if (raw) {
    return {
      headers: response.headers,
      stream,
    };
  } else {
    return stream;
  }
}

async function fetchData(url: string, options: Partial<FetchOptions> = {}, client: AxiosInstance = axios) {
  const { raw, ...rest } = options;
  const response = await _fetch(url, { responseType: "json", ...rest }, client);
  const data = response.data;

  if (raw) {
    return {
      headers: response.headers,
      data,
    };
  } else {
    return data;
  }
}

async function fetchJson(url: string, options: Partial<FetchOptions> = {}) {
  const response = await _fetch(url, { ...options, responseType: "json" });
  return response.data;
}

function addCsvHeaders(filename: string, encoding: string, res: any) {
  res.setHeader("Content-disposition", `attachment; filename=${filename}`);
  res.setHeader("Content-Type", `text/csv; charset=${encoding}`);
}

const getHttpClient = (options: CreateAxiosDefaults<any> = {}) =>
  axios.create({
    timeout: 15000,
    httpAgent: new http.Agent({ keepAlive: true }),
    httpsAgent: new https.Agent({ keepAlive: true }),
    ...options,
  });

export { addCsvHeaders, fetchJson, fetchStream, fetchData, getHttpClient };
