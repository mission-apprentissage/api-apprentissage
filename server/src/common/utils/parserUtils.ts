import { Options, parse } from "csv-parse";
import { isEmpty, pickBy } from "lodash-es";

export const DEFAULT_DELIMITER = ";";

export function parseCsv(options: Options = {}) {
  return parse({
    trim: true,
    delimiter: DEFAULT_DELIMITER,
    columns: true,
    bom: true,
    relax_column_count: true,
    on_record: (record) => {
      return pickBy(record, (v) => {
        return !isEmpty(v) && v.trim().length;
      });
    },
    ...options,
  });
}
