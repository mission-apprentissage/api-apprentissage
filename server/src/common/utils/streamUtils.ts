import { Transform } from "node:stream";

type Options = {
  size: number;
};

export function createBatchTransformStream(opts: Options): Transform {
  let currentBatch: unknown[] = [];

  return new Transform({
    objectMode: true,
    transform(chunk, _encoding, callback) {
      currentBatch.push(chunk);

      if (currentBatch.length >= opts.size) {
        this.push(currentBatch);
        currentBatch = [];
      }

      callback();
    },
    flush(callback) {
      if (currentBatch.length > 0) {
        this.push(currentBatch);
      }
      callback();
    },
  });
}
