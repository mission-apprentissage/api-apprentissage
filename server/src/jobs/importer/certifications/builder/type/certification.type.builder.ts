import { internal } from "@hapi/boom";
import { ICertification } from "api-alternance-sdk";

import { ISourceAggregatedData } from "@/jobs/importer/certifications/builder/certification.builder.js";

export function buildCertificationType(data: ISourceAggregatedData): ICertification["type"] {
  const cfdData = data.bcn?.data ?? null;
  const rncpData = data.france_competence?.data ?? null;

  return {
    certificateurs_rncp:
      rncpData?.certificateurs.map((certificateur) => ({
        siret: certificateur.Siret_Certificateur,
        nom: certificateur.Nom_Certificateur,
      })) ?? null,
    enregistrement_rncp: rncpData?.standard?.Type_Enregistrement ?? null,
    gestionnaire_diplome: cfdData?.GESTIONNAIRE_FORMATION_DIPLOME ?? null,
    nature: {
      cfd:
        cfdData === null
          ? null
          : {
              code: cfdData.NATURE_FORMATION_DIPLOME ?? null,
              libelle: cfdData.N_NATURE_FORMATION_DIPLOME_LIBELLE_100 ?? null,
            },
    },
    voie_acces: {
      rncp:
        rncpData === null
          ? null
          : rncpData.voies_d_acces.reduce(
              (acc, voie) => {
                if (voie.Si_Jury === "En contrat d’apprentissage") {
                  acc.apprentissage = true;
                } else if (voie.Si_Jury === "Par expérience") {
                  acc.experience = true;
                } else if (voie.Si_Jury === "Par candidature individuelle") {
                  acc.candidature_individuelle = true;
                } else if (voie.Si_Jury === "En contrat de professionnalisation") {
                  acc.contrat_professionnalisation = true;
                } else if (voie.Si_Jury === "Après un parcours de formation continue") {
                  acc.formation_continue = true;
                } else if (voie.Si_Jury === "Après un parcours de formation sous statut d’élève ou d’étudiant") {
                  acc.formation_statut_eleve = true;
                } else {
                  throw internal("import.certifications: unexpected voie d'acces value", { voie });
                }

                return acc;
              },
              {
                apprentissage: false as boolean,
                experience: false as boolean,
                candidature_individuelle: false as boolean,
                contrat_professionnalisation: false as boolean,
                formation_continue: false as boolean,
                formation_statut_eleve: false as boolean,
              }
            ),
    },
  };
}
