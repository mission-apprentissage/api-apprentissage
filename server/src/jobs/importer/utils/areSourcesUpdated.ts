export function areSourcesUpdated(
  lastImportSources: Record<string, { import_date: Date }> | null | undefined,
  currentSources: Record<string, { import_date: Date }>
): boolean {
  if (!lastImportSources) {
    return true;
  }

  for (const source in currentSources) {
    if (!lastImportSources[source] || lastImportSources[source].import_date < currentSources[source].import_date) {
      return true;
    }
  }

  return false;
}
