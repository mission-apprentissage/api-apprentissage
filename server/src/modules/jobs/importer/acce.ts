import type { ReadStream } from "node:fs";
import { Transform } from "node:stream";
import { pipeline } from "node:stream/promises";

import { internal } from "@hapi/boom";
import { parse } from "csv-parse";
import { ObjectId } from "mongodb";
import { IAcce, ZAcceByType } from "shared/models/acce/acce.model";
import unzipper from "unzipper";

import parentLogger from "@/common/logger";

import { downloadCsvExtraction } from "../../../common/apis/acce";
import { withCause } from "../../../common/errors/withCause";
import { getDbCollection } from "../../../common/utils/mongodbUtils";
import { createBatchTransformStream } from "../../../common/utils/streamUtils";

const logger = parentLogger.child({ module: "import:acce" });

async function parseAcceFile(stream: ReadStream, source: string, date: Date) {
  try {
    const zod = ZAcceByType[source as keyof typeof ZAcceByType] ?? null;

    if (zod === null) {
      throw internal("import.acce: unexpected source file", { source });
    }

    await pipeline(
      stream,
      parse({
        bom: true,
        columns: true,
        relax_column_count: true,
        encoding: "latin1",
        delimiter: ";",
        trim: true,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        onRecord: (record, { columns }: any) => {
          const data = columns.reduce((acc: Record<string, string | null>, column: { name: string }) => {
            acc[column.name] = record[column.name]?.trim() || null;
            return acc;
          }, {});

          return zod.parse({
            _id: new ObjectId(),
            source,
            date,
            data,
          });
        },
      }),
      createBatchTransformStream({ size: 1_000 }),
      new Transform({
        objectMode: true,
        async transform(chunk, _encoding, callback) {
          try {
            await getDbCollection("source.acce").insertMany(chunk);
            callback();
          } catch (error) {
            callback(withCause(internal("import.acce: error when inserting"), error));
          }
        },
      })
    );

    await getDbCollection("source.acce").deleteMany({
      source: source as IAcce["source"],
      date: { $ne: date },
    });
  } catch (error) {
    throw withCause(internal("import.acce: unable to parseAcceFile", { source }), error);
  }
}

export async function importAcceData(readStream: ReadStream) {
  const importDate = new Date();

  const zip = readStream.pipe(unzipper.Parse({ forceStream: true }));
  for await (const entry of zip) {
    await parseAcceFile(entry, entry.path, importDate);
    entry.autodrain();
  }
}

export async function runAcceImporter() {
  logger.info("Geting ACCE file...");

  const stream = await downloadCsvExtraction();

  logger.info("Import ACCE data starting...");

  await importAcceData(stream);
}
