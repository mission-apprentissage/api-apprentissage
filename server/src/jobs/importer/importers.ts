import { getAcceeImporterStatus, runAcceImporter } from "./acce/acce.js";
import { getBcnImporterStatus, runBcnImporter } from "./bcn/bcn.importer.js";
import { getCatalogueImporterStatus, runCatalogueImporter } from "./catalogue/catalogue.importer.js";
import { getCertificationImporterStatus, importCertifications } from "./certifications/certifications.importer.js";
import { getCommuneImporterStatus, runCommuneImporter } from "./commune/commune.importer.js";
import { getDaresApiIdccImporterStatus, runDaresApeIdccImporter } from "./dares/ape_idcc/dares.ape_idcc.importer.js";
import { getDaresCcnImporterStatus, runDaresConventionCollectivesImporter } from "./dares/ccn/dares.ccn.importer.js";
import { getFormationsImporterStatus, importFormations } from "./formation/formation.importer.js";
import { getFranceCompetencesImporterStatus, runRncpImporter } from "./france_competence/france_competence.importer.js";
import { getKaliImporterStatus, runKaliConventionCollectivesImporter } from "./kali/kali.ccn.importer.js";
import { getKitApprentissageImporterStatus, runKitApprentissageImporter } from "./kit/kitApprentissage.importer.js";
import { getMissionLocaleImporterStatus, runMissionLocaleImporter } from "./mission_locale/mission_locale.importer.js";
import { getOrganismesImporterStatus, importOrganismes } from "./organisme/organisme.importer.js";
import { getReferentielImporterStatus, runReferentielImporter } from "./referentiel/referentiel.js";
import type { Importer } from "./types.js";

const timings = {
  import_source: "0 4 * * *",
  import_source_main: "0 */4 * * *",
};

export const importers: Record<string, Importer> = {
  "Mise à jour acce": {
    cron_string: timings.import_source,
    handler: runAcceImporter,
    resumable: true,
    getStatus: getAcceeImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données BCN": {
    cron_string: timings.import_source_main,
    handler: runBcnImporter,
    resumable: true,
    getStatus: getBcnImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Kit Apprentissage": {
    cron_string: timings.import_source_main,
    handler: runKitApprentissageImporter,
    resumable: true,
    getStatus: getKitApprentissageImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Referentiel": {
    cron_string: timings.import_source_main,
    handler: runReferentielImporter,
    resumable: true,
    getStatus: getReferentielImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données Catalogue": {
    cron_string: timings.import_source_main,
    handler: runCatalogueImporter,
    resumable: true,
    getStatus: getCatalogueImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des données France Compétences": {
    cron_string: timings.import_source,
    handler: runRncpImporter,
    resumable: true,
    getStatus: getFranceCompetencesImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des certifications": {
    cron_string: timings.import_source_main,
    handler: async () => importCertifications(),
    resumable: true,
    getStatus: getCertificationImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des organismes": {
    cron_string: timings.import_source_main,
    handler: async () => importOrganismes(false),
    resumable: true,
    getStatus: getOrganismesImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des formations": {
    cron_string: timings.import_source_main,
    handler: importFormations,
    resumable: true,
    getStatus: getFormationsImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Conventions Collective Kali": {
    cron_string: timings.import_source,
    handler: runKaliConventionCollectivesImporter,
    resumable: true,
    getStatus: getKaliImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Conventions Collective Dares": {
    cron_string: timings.import_source,
    handler: runDaresConventionCollectivesImporter,
    resumable: true,
    getStatus: getDaresCcnImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des APE-IDCC Dares": {
    cron_string: timings.import_source,
    handler: runDaresApeIdccImporter,
    resumable: true,
    getStatus: getDaresApiIdccImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Missions Locales": {
    cron_string: timings.import_source,
    handler: runMissionLocaleImporter,
    resumable: true,
    getStatus: getMissionLocaleImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
  "Import des Communes": {
    cron_string: timings.import_source,
    handler: runCommuneImporter,
    resumable: true,
    getStatus: getCommuneImporterStatus,
    checkinMargin: 60, // 1h
    maxRuntimeInMinutes: 30,
  },
};
