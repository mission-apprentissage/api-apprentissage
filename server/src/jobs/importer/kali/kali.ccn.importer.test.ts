import { useMongo } from "@tests/mongo.test.utils.js";
import { createReadStream } from "fs";
import { ObjectId } from "mongodb";
import nock, { cleanAll, disableNetConnect, enableNetConnect } from "nock";
import { dirname, join } from "path";
import type { IDataGouvDataset } from "shared";
import type { IImportMeta } from "shared/models/import.meta.model";
import { fileURLToPath } from "url";
import { beforeEach, describe, expect, it, vi } from "vitest";

import { fetchDataGouvDataSet } from "@/services/apis/data_gouv/data_gouv.api.js";
import { getDbCollection } from "@/services/mongodb/mongodbService.js";

import { runKaliConventionCollectivesImporter } from "./kali.ccn.importer.js";

vi.mock("@/services/apis/data_gouv/data_gouv.api", async (importOriginal) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const mod = (await importOriginal()) as any;
  return {
    ...mod,
    fetchDataGouvDataSet: vi.fn(),
  };
});

describe("runConventionCollectivesImporter", () => {
  useMongo();

  const now = new Date("2024-06-14T09:00:07.000Z");

  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(now);
    disableNetConnect();

    return () => {
      vi.mocked(fetchDataGouvDataSet).mockReset();
      vi.useRealTimers();
      cleanAll();
      enableNetConnect();
    };
  });

  it("should import ccn data", async () => {
    const ccnResource = {
      created_at: new Date("2023-05-02T16:44:08.913000+00:00"),
      id: "02b67492-5243-44e8-8dd1-0cb3f90f35ff",
      last_modified: new Date("2023-03-06T15:14:03+00:00"),
      latest: "https://www.data.gouv.fr/fr/datasets/r/02b67492-5243-44e8-8dd1-0cb3f90f35ff",
      title: "Liste des conventions collectives",
    };
    const otherResource = {
      created_at: new Date("2017-09-04T16:34:48.090000+00:00"),
      id: "c35c7767-335f-4f25-80d8-2d327a30a922",
      last_modified: new Date("2017-08-28T16:54:05+00:00"),
      latest: "https://www.data.gouv.fr/fr/datasets/r/c35c7767-335f-4f25-80d8-2d327a30a922",
      title: "Présentation de KALI",
    };

    const dataset: IDataGouvDataset = {
      id: "53ba5033a3a729219b7bead9",
      title: "KALI : Conventions collectives nationales",
      resources: [ccnResource, otherResource],
    };

    vi.mocked(fetchDataGouvDataSet).mockResolvedValue(dataset);

    const dataFixture = join(dirname(fileURLToPath(import.meta.url)), "fixtures/sample.xlsx");

    nock("https://www.data.gouv.fr/fr")
      .get(`/datasets/r/02b67492-5243-44e8-8dd1-0cb3f90f35ff`)
      .reply(200, createReadStream(dataFixture));

    const initialImports: IImportMeta[] = [
      {
        _id: new ObjectId(),
        import_date: new Date("2024-06-13T09:00:07.000Z"),
        type: "kali_ccn",
        status: "done",
      },
    ];

    await getDbCollection("import.meta").insertMany(initialImports);
    await getDbCollection("source.kali.ccn").insertMany([
      {
        _id: new ObjectId("666c3f138a5c98c03c14e94b"),
        date_import: new Date("2024-06-14T13:01:06.644Z"),
        data: {
          type: "TI",
          id: "KALICONT000046660537",
          titre:
            "Accord professionnel du 21 juillet 2022 relatif à la reconversion ou la promotion par l'alternance « Pro-A »",
          nature: "ACCORD PROFESSIONNEL",
          etat: "VIGUEUR_ETEN",
          debut: new Date("2022-11-25T23:00:00.000Z"),
          fin: null,
          url: "https://www.legifrance.gouv.fr/conv_coll/id/KALICONT000046660537",
        },
      },
      {
        _id: new ObjectId("666c3f138a5c98c03c14e957"),
        date_import: new Date("2024-06-14T13:01:06.644Z"),
        data: {
          type: "IDCC",
          id: "KALICONT000044617302",
          idcc: 3237,
          titre:
            "Convention collective nationale des métiers du commerce de détail alimentaire spécialisé du 12 janvier 2021 - Etendue par arrêté du 17 septembre 2021 JORF 23 décembre 2021",
          nature: "CONVENTION COLLECTIVE NATIONALE",
          etat: "VIGUEUR_ETEN",
          debut: new Date("2021-12-31T23:00:00.000Z"),
          fin: null,
          url: "https://www.legifrance.gouv.fr/conv_coll/id/KALICONT000044617302",
        },
      },
    ]);

    await runKaliConventionCollectivesImporter();

    await expect(getDbCollection("import.meta").find({}).toArray()).resolves.toEqual([
      ...initialImports,
      {
        _id: expect.any(ObjectId),
        import_date: now,
        type: "kali_ccn",
        status: "done",
      },
    ]);

    const result = await getDbCollection("source.kali.ccn")
      .find({}, { projection: { _id: 0 } })
      .toArray();
    expect(result).toMatchSnapshot();
  });
});
