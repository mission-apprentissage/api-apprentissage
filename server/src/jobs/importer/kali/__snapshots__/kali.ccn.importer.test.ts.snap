// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runConventionCollectivesImporter > should import ccn data 1`] = `
[
  {
    "data": {
      "debut": 2022-11-25T23:00:00.000Z,
      "etat": "VIGUEUR_ETEN",
      "fin": null,
      "id": "KALICONT000046660537",
      "nature": "ACCORD PROFESSIONNEL",
      "titre": "Accord professionnel du 21 juillet 2022 relatif à la reconversion ou la promotion par l'alternance « Pro-A »",
      "type": "TI",
      "url": "https://www.legifrance.gouv.fr/conv_coll/id/KALICONT000046660537",
    },
    "date_import": 2024-06-14T13:01:06.644Z,
  },
  {
    "data": {
      "debut": 2021-12-31T23:00:00.000Z,
      "etat": "VIGUEUR_ETEN",
      "fin": null,
      "id": "KALICONT000044617302",
      "idcc": 3237,
      "nature": "CONVENTION COLLECTIVE NATIONALE",
      "titre": "Convention collective nationale des métiers du commerce de détail alimentaire spécialisé du 12 janvier 2021 - Etendue par arrêté du 17 septembre 2021 JORF 23 décembre 2021",
      "type": "IDCC",
      "url": "https://www.legifrance.gouv.fr/conv_coll/id/KALICONT000044617302",
    },
    "date_import": 2024-06-14T13:01:06.644Z,
  },
  {
    "data": {
      "debut": 2023-12-31T23:00:00.000Z,
      "etat": "VIGUEUR_ETEN",
      "fin": null,
      "id": "KALICONT000046993250",
      "idcc": 3248,
      "nature": "CONVENTION COLLECTIVE NATIONALE",
      "titre": "Convention collective nationale de la métallurgie du 7 février 2022 - Étendue par arrêté du 14 décembre 2022 JORF 22 décembre 2022",
      "type": "IDCC",
      "url": "https://www.legifrance.gouv.fr/conv_coll/id/KALICONT000046993250",
    },
    "date_import": 2024-06-14T09:00:07.000Z,
  },
  {
    "data": {
      "debut": 2023-02-11T23:00:00.000Z,
      "etat": "VIGUEUR_ETEN",
      "fin": null,
      "id": "KALICONT000047124320",
      "nature": "ACCORD",
      "titre": "Accord du 7 octobre 2022 relatif à la gestion prévisionnelle des emplois et des compétences des salariés permanents",
      "type": "TI",
      "url": "https://www.legifrance.gouv.fr/conv_coll/id/KALICONT000047124320",
    },
    "date_import": 2024-06-14T09:00:07.000Z,
  },
]
`;
